src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,210 @@
+        }
+
+        private async Task DownloadFileContainerAsync(Guid projectId, BuildArtifact artifact, string rootPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        {
+            var containerIdAndRoot = ParseContainerId(artifact.Resource.Data);
+
+            var items = await containerClient.QueryContainerItemsAsync(
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, tracer);
+            if (minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs);
+            }
+
+            var folderItems = items.Where(i => i.ItemType == ContainerItemType.Folder);
+            if (folderItems != null)
+            {
+                Parallel.ForEach(folderItems, (folder) =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, folder, artifact.Name);
+                    Directory.CreateDirectory(targetPath);
+                });
+            }
+
+            var fileItems = items.Where(i => i.ItemType == ContainerItemType.File);
+
+            var batchItemsBlock = new BatchBlock<FileContainerItem>(
+                batchSize: 1000,
+                new GroupingDataflowBlockOptions()
+                {
+                    CancellationToken = cancellationToken,
+                });
+
+            var fetchStream = NonSwallowingTransformManyBlock.Create<IEnumerable<FileContainerItem>, (Stream, string)>(
+                async itemsStream =>
+                {
+                    List<(Stream, string)> collection= new List<(Stream, string)>();
+                    foreach (var item in itemsStream)
+                    {
+                        var targetPath = ResolveTargetPath(rootPath, item, artifact.Name);
+                        Stream stream = await this.DownloadFileFromContainerAsync(containerIdAndRoot, projectId, containerClient, item, targetPath, cancellationToken);
+                        collection.Add((stream, targetPath));
+                    }
+                    return collection;
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 1000,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount * 8,
+                    CancellationToken = cancellationToken,
+                });
+
+            var downloadBlock = NonSwallowingActionBlock.Create<(Stream stream, string targetPath)>(
+                item =>
+                {
+                    using (item.stream)
+                    {
+                        int retryCount = 0;
+                        try
+                        {
+                            using (var fileStream = new FileStream(item.targetPath, FileMode.Create))
```

Comment:
>FileMode.Create [](start = 84, length = 15)

this doesn't overwrite?

Reply:
Actually, I think it does - "Specifies that the operating system should create a new file. If the file already
        //     exists, it will be overwritten.: